pkg/steps/artifacts.go

```diff
@@ -207,43 +208,66 @@ func (c podClient) WithNewLoggingClient() PodClient {
 	}
 }
 
-func waitForContainer(podClient PodClient, ns, name, containerName string) error {
-	logrus.WithFields(logrus.Fields{
-		"namespace": ns,
-		"name":      name,
-		"container": containerName,
-	}).Trace("Waiting for container to be running.")
-
-	ctx := context.TODO()
+type evaluator func(obj runtime.Object) (bool, error)
 
+// waitForConditionOnObject uses a watch to wait for a condition to be true on an object.
+// When the condition is satisfied or the timeout expires, the object is returned along
+// with any errors encountered.
+func waitForConditionOnObject(ctx context.Context, client watchingclient.Client, identifier ctrlruntimeclient.ObjectKey, list ctrlruntimeclient.ObjectList, into ctrlruntimeclient.Object, evaluate evaluator, timeout time.Duration) error {
 	lw := &cache.ListWatch{
 		ListFunc: func(options metav1.ListOptions) (object runtime.Object, e error) {
-			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", name).String()
-			res := &corev1.PodList{}
-			return res, podClient.List(ctx, res, &ctrlruntimeclient.ListOptions{Namespace: ns, Raw: &options})
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", identifier.Name).String()
+			res := list
+			return res, client.List(ctx, res, &ctrlruntimeclient.ListOptions{Namespace: identifier.Namespace, Raw: &options})
 		},
 		WatchFunc: func(options metav1.ListOptions) (i watch.Interface, e error) {
 			opts := &ctrlruntimeclient.ListOptions{
-				Namespace:     ns,
-				FieldSelector: fields.OneTermEqualSelector("metadata.name", name),
+				Namespace:     identifier.Namespace,
+				FieldSelector: fields.OneTermEqualSelector("metadata.name", identifier.Name),
 				Raw:           &options,
 			}
-			return podClient.Watch(ctx, &corev1.PodList{}, opts)
+			res := list
+			return client.Watch(ctx, res, opts)
 		},
 	}
 
 	// pods in this call here are always expected to exist in advance
-	var podExistsPrecondition toolswatch.PreconditionFunc
+	var existsPrecondition toolswatch.PreconditionFunc
 
 	waitForContainerStatus := func(event watch.Event) (bool, error) {
 		if event.Type == watch.Deleted {
-			return false, fmt.Errorf("pod/%v ns/%v was deleted", name, ns)
+			return false, fmt.Errorf("%s was deleted", identifier.String())
 		}
 		// the outer library will handle errors, we have no pod data to review in this case
 		if event.Type != watch.Added && event.Type != watch.Modified {
 			return false, nil
 		}
-		switch pod := event.Object.(type) {
+		return evaluate(event.Object)
+	}
+
+	waitTimeout, cancel := toolswatch.ContextWithOptionalTimeout(ctx, timeout /*weird, but this is what it was*/)
+	defer cancel()
+	_, syncErr := toolswatch.UntilWithSync(waitTimeout, lw, &corev1.Pod{}, existsPrecondition, waitForContainerStatus)
```

Comment:
You cannot use `&corev1.Pod{}` here, this is a generic function.

Reply:
Updated.